server/interview/algorithm.py

```diff
@@ -1,84 +1,193 @@
+        """
+        pair members based on common availability slots. Number of members must be even.
+        """
+        self._validate_input(pool_member_ids)
+
+        if not self._availabilities:
+            raise ValueError("Availabilities must be set before pairing")
+
+        if not all(mid in self._availabilities for mid in pool_member_ids):
+            raise ValueError("All pool members must have availabilities")
+
+        common_slots = self._calculate_common_slots_matrix(pool_member_ids)
+        preferences = self._calculate_preferences(pool_member_ids, common_slots)
+        pairs = self._stable_matching(preferences)
+
+        return MatchingResult(
+            pairs=pairs, common_slots=common_slots, preference_scores=preferences
+        )
+
+    def _calculate_common_slots_matrix(
+        self, pool_member_ids: List[int]
+    ) -> NDArray[np.int_]:
+        num_members = len(pool_member_ids)
+        common_slots = np.zeros((num_members, num_members), dtype=np.int_)
+
+        for i in range(num_members):
+            availability1 = self._availabilities[pool_member_ids[i]]
+
+            for j in range(i + 1, num_members):
+                availability2 = self._availabilities[pool_member_ids[j]]
+                common_count = self.calculate_common_slots_numpy(
+                    availability1, availability2
+                )
+                common_slots[i, j] = common_count
+                common_slots[j, i] = common_count
+
+        return common_slots
+
+    def calculate_preferences(
+        self, pool_member_ids: List[int], availabilities: Dict[int, List[List[bool]]]
+    ) -> Dict[int, List[Tuple[int, int]]]:
+        """
+        calculate preference lists for all members based on common availability.
+        """
+        self.set_availabilities(availabilities)
+        self._validate_input(pool_member_ids, require_even=False)
+
+        common_slots = self._calculate_common_slots_matrix(pool_member_ids)
+        return self._calculate_preferences(pool_member_ids, common_slots)
+
+    def _calculate_preferences(
+        self, pool_member_ids: List[int], common_slots: NDArray[np.int_]
+    ) -> Dict[int, List[Tuple[int, int]]]:
+        num_members = len(pool_member_ids)
+        preferences: Dict[int, List[Tuple[int, int]]] = {}
+
+        for i in range(num_members):
+            prefs = [(j, common_slots[i, j]) for j in range(num_members) if i != j]
+            preferences[i] = sorted(prefs, key=lambda x: (x[1], -x[0]), reverse=True)
+
+        return preferences
+
+    def _stable_matching(
+        self, preferences: Dict[int, List[Tuple[int, int]]]
+    ) -> List[int]:
+        """
+        Gale-Shapley algorithm for stable matching.
+        """
+        n = len(preferences)
         if n < 2:
             raise ValueError("At least two members are required for matching")
-        if n == 2:
-            return [1, 0]
 
         free_members = list(range(n))
```

Comment:
Maybe use a [`deque`](https://docs.python.org/3/library/collections.html#collections.deque) here instead since we're repeatedly popping off from the front of this list.

Reply:
Surprisingly...

```
before (with list):
Time taken for pairing 20 members: 2.02 ms
Time taken for pairing 50 members: 12.17 ms
Time taken for pairing 100 members: 47.39 ms
Time taken for pairing 200 members: 194.36 ms
Time taken for pairing 500 members: 1263.41 ms
Time taken for pairing 1000 members: 5144.68 ms
Time taken for pairing 2000 members: 20620.16 ms
```

```
after (with deque):
Time taken for pairing 20 members: 1.99 ms
Time taken for pairing 50 members: 11.89 ms
Time taken for pairing 100 members: 48.04 ms
Time taken for pairing 200 members: 193.21 ms
Time taken for pairing 500 members: 1294.98 ms
Time taken for pairing 1000 members: 5157.06 ms
Time taken for pairing 2000 members: 20855.78 ms
```

will still change to deque though